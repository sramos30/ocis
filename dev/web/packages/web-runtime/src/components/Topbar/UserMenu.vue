<template>
  <nav v-if="userId" :aria-label="$gettext('Account menu')">
    <oc-button
      id="_userMenuButton"
      ref="menuButton"
      class="oc-topbar-personal"
      appearance="raw"
      :aria-label="$gettext('User Menu')"
    >
      <avatar-image
        class="oc-topbar-personal-avatar oc-flex-inline oc-flex-center oc-flex-middle"
        :width="32"
        :userid="userId"
        :user-name="user.displayname"
      />
    </oc-button>
    <oc-drop
      ref="menu"
      drop-id="account-info-container"
      toggle="#_userMenuButton"
      mode="click"
      close-on-click
      padding-size="small"
    >
      <oc-list class="user-menu-list">
        <li>
          <oc-button
            id="oc-topbar-account-manage"
            type="router-link"
            :to="{ path: '/account' }"
            appearance="raw"
          >
            <avatar-image :width="32" :userid="userId" :user-name="user.displayname" />
            <span class="profile-info-wrapper" :class="{ 'oc-py-xs': !user.email }">
              <span v-text="user.displayname" />
              <br v-if="user.email" />
              <span v-if="user.email" class="oc-text-small" v-text="user.email" />
            </span>
          </oc-button>
        </li>
        <li v-for="(app, index) in applicationsList" :key="`user-menu-${index}`">
          <oc-button v-if="app.url" type="a" appearance="raw" :target="app.target" :href="app.url">
            <oc-icon :name="app.icon" class="oc-p-xs" />
            <translate>{{ app.title }}</translate>
          </oc-button>
          <oc-button v-else type="router-link" appearance="raw" :to="{ path: app.path }">
            <oc-icon :name="app.icon" class="oc-p-xs" />
            <translate>{{ app.title }}</translate>
          </oc-button>
        </li>
        <li>
          <oc-button id="oc-topbar-account-logout" appearance="raw" @click="logout">
            <oc-icon name="logout-box-r" fill-type="line" class="oc-p-xs" />
            <translate>Log out</translate>
          </oc-button>
        </li>
        <li v-if="quotaEnabled" class="storage-wrapper oc-pl-s">
          <oc-icon name="cloud" fill-type="line" class="oc-p-xs" />
          <div class="storage-wrapper-text">
            <p class="oc-my-rm">
              <span v-text="personalStorageLabel" />
              <br />
              <span class="oc-text-small" v-text="personalStorageDetailsLabel" />
            </p>
            <oc-progress
              :value="parseInt(quotaUsagePercent)"
              :max="100"
              size="small"
              :variation="quotaProgressVariant"
            />
          </div>
        </li>
      </oc-list>
    </oc-drop>
  </nav>
</template>

<script>
import { mapGetters } from 'vuex'
import filesize from 'filesize'

export default {
  props: {
    applicationsList: {
      type: Array,
      required: false,
      default: () => []
    }
  },
  computed: {
    ...mapGetters(['quota', 'user']),

    userId() {
      return this.user.username || this.user.id
    },
    personalStorageLabel() {
      return this.$gettextInterpolate(this.$gettext('Personal storage (%{percentage}% used)'), {
        percentage: this.quotaUsagePercent
      })
    },
    personalStorageDetailsLabel() {
      return this.$gettextInterpolate(this.$gettext('%{used} of %{total} used'), {
        used: this.quotaUsed,
        total: this.quotaTotal
      })
    },
    quotaEnabled() {
      return !!this.quota
    },
    quotaTotal() {
      return filesize(this.quota.total)
    },
    quotaUsed() {
      return filesize(this.quota.used)
    },
    quotaUsagePercent() {
      return ((this.quota.used / this.quota.total) * 100).toFixed(1)
    },
    quotaProgressVariant() {
      if (this.quotaUsagePercent < 80) return 'primary'
      if (this.quotaUsagePercent < 90) return 'warning'
      return 'danger'
    }
  },
  mounted() {
    this.$refs.menu?.tippy?.setProps({
      onHidden: () => this.$refs.menuButton.$el.focus(),
      onShown: () => this.$refs.menu.$el.querySelector('a:first-of-type').focus()
    })
  },
  methods: {
    logout() {
      // Use timeout to leave enough time for the dropdown to be hidden
      setTimeout(() => {
        this.$store.dispatch('logout')
      })
    }
  }
}
</script>

<style lang="scss" scoped>
.user-menu-list li {
  align-items: center;
  display: flex;
  margin: var(--oc-space-xsmall) 0;

  a,
  button {
    gap: var(--oc-space-medium);
    justify-content: flex-start;
    min-height: 3rem;
    padding-left: var(--oc-space-small);
    width: 100%;

    &:focus,
    &:hover {
      background-color: var(--oc-color-background-hover);
      color: var(--oc-color-swatch-passive-default);
      text-decoration: none;
    }

    .profile-info-wrapper {
      text-align: left;
    }
  }

  &.storage-wrapper {
    gap: var(--oc-space-medium);
    min-height: 3rem;

    .storage-wrapper-text {
      align-self: flex-end;
      display: inline-block;
    }
  }
}
</style>
