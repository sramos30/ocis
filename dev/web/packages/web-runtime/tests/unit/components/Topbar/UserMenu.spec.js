import { createLocalVue, mount } from '@vue/test-utils'
import Vuex from 'vuex'
import DesignSystem from 'owncloud-design-system'
import GetTextPlugin from 'vue-gettext'
import UserMenu from 'web-runtime/src/components/Topbar/UserMenu.vue'
import stubs from '../../../../../../tests/unit/stubs'

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(DesignSystem)
localVue.use(GetTextPlugin, {
  translations: 'does-not-matter.json',
  silent: true
})

const basicQuota = 300
const warningQuota = 810
const dangerQuota = 910

const noEmail = ''
const email = 'test@test.de'

describe('User Menu component', () => {
  describe('when quota and no email is set', () => {
    it('renders a navigation without email', () => {
      const wrapper = getMountedWrapper({ used: basicQuota, total: 1000 }, noEmail)
      expect(wrapper).toMatchSnapshot()
    })
  })
  describe('when no quota and email is set', () => {
    it('the user menu does not contain a quota', () => {
      const wrapper = getMountedWrapper(null, email)
      expect(wrapper).toMatchSnapshot()
    })
  })
  describe('when no quota and no email is set', () => {
    it('the user menu does not contain a quota', () => {
      const wrapper = getMountedWrapper(null, noEmail)
      expect(wrapper).toMatchSnapshot()
    })
  })
  describe('when quota is below 80%', () => {
    it('renders a primary quota progress bar', () => {
      const wrapper = getMountedWrapper({ used: basicQuota, total: 1000 }, email)
      expect(wrapper).toMatchSnapshot()
    })
  })
  describe('when quota is above 80% and below 90%', () => {
    it('renders a warning quota progress bar', () => {
      const wrapper = getMountedWrapper({ used: warningQuota, total: 1000 }, email)
      expect(wrapper).toMatchSnapshot()
    })
  })
  describe('when basic quota is set', () => {
    it('renders a danger quota progress bar', () => {
      const wrapper = getMountedWrapper({ used: dangerQuota, total: 1000 }, email)
      expect(wrapper).toMatchSnapshot()
    })
  })
})

const getMountedWrapper = (quota, userEmail) => {
  return mount(UserMenu, {
    store: new Vuex.Store({
      getters: {
        quota: () => quota,
        user: () => ({
          id: 'einstein',
          username: 'einstein',
          userDisplayName: 'Albert Einstein',
          userEmail
        })
      }
    }),
    localVue,
    propsData: {
      applicationsList: [
        {
          icon: 'application',
          path: '/settings',
          title: 'Settings'
        }
      ]
    },
    stubs
  })
}
