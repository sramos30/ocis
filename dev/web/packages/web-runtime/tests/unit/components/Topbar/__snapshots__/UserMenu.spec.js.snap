// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Menu component when basic quota is set renders a danger quota progress bar 1`] = `
<nav aria-label="Account menu">
  <oc-button-stub type="button" size="medium" arialabel="User Menu" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium" id="_userMenuButton" class="oc-topbar-personal">
    <avatar-image-stub width="32" userid="einstein" class="oc-topbar-personal-avatar oc-flex-inline oc-flex-center oc-flex-middle"></avatar-image-stub>
  </oc-button-stub>
  <oc-drop-stub dropid="account-info-container" toggle="#_userMenuButton" position="bottom-start" mode="click" closeonclick="true" paddingsize="small">
    <oc-list-stub class="user-menu-list">
      <li>
        <oc-button-stub type="router-link" size="medium" to="[object Object]" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium" id="oc-topbar-account-manage">
          <avatar-image-stub width="32" userid="einstein"></avatar-image-stub> <span class="profile-info-wrapper oc-py-xs"><span></span>
          <!---->
          <!----></span>
        </oc-button-stub>
      </li>
      <li>
        <oc-button-stub type="router-link" size="medium" to="[object Object]" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium">
          <oc-icon-stub name="application" filltype="fill" accessiblelabel="" type="span" size="medium" variation="passive" color="" class="oc-p-xs"></oc-icon-stub>
          <translate-stub tag="span">Settings</translate-stub>
        </oc-button-stub>
      </li>
      <li>
        <oc-button-stub type="button" size="medium" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium" id="oc-topbar-account-logout">
          <oc-icon-stub name="logout-box-r" filltype="line" accessiblelabel="" type="span" size="medium" variation="passive" color="" class="oc-p-xs"></oc-icon-stub>
          <translate-stub tag="span">Log out</translate-stub>
        </oc-button-stub>
      </li>
      <li class="storage-wrapper oc-pl-s">
        <oc-icon-stub name="cloud" filltype="line" accessiblelabel="" type="span" size="medium" variation="passive" color="" class="oc-p-xs"></oc-icon-stub>
        <div class="storage-wrapper-text">
          <p class="oc-my-rm"><span>Personal storage (91.0% used)</span> <br> <span class="oc-text-small">910 B of 1 kB used</span></p>
          <oc-progress-stub value="91" max="100" size="small" variation="danger"></oc-progress-stub>
        </div>
      </li>
    </oc-list-stub>
  </oc-drop-stub>
</nav>
`;

exports[`User Menu component when no quota and email is set the user menu does not contain a quota 1`] = `
<nav aria-label="Account menu">
  <oc-button-stub type="button" size="medium" arialabel="User Menu" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium" id="_userMenuButton" class="oc-topbar-personal">
    <avatar-image-stub width="32" userid="einstein" class="oc-topbar-personal-avatar oc-flex-inline oc-flex-center oc-flex-middle"></avatar-image-stub>
  </oc-button-stub>
  <oc-drop-stub dropid="account-info-container" toggle="#_userMenuButton" position="bottom-start" mode="click" closeonclick="true" paddingsize="small">
    <oc-list-stub class="user-menu-list">
      <li>
        <oc-button-stub type="router-link" size="medium" to="[object Object]" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium" id="oc-topbar-account-manage">
          <avatar-image-stub width="32" userid="einstein"></avatar-image-stub> <span class="profile-info-wrapper oc-py-xs"><span></span>
          <!---->
          <!----></span>
        </oc-button-stub>
      </li>
      <li>
        <oc-button-stub type="router-link" size="medium" to="[object Object]" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium">
          <oc-icon-stub name="application" filltype="fill" accessiblelabel="" type="span" size="medium" variation="passive" color="" class="oc-p-xs"></oc-icon-stub>
          <translate-stub tag="span">Settings</translate-stub>
        </oc-button-stub>
      </li>
      <li>
        <oc-button-stub type="button" size="medium" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium" id="oc-topbar-account-logout">
          <oc-icon-stub name="logout-box-r" filltype="line" accessiblelabel="" type="span" size="medium" variation="passive" color="" class="oc-p-xs"></oc-icon-stub>
          <translate-stub tag="span">Log out</translate-stub>
        </oc-button-stub>
      </li>
      <!---->
    </oc-list-stub>
  </oc-drop-stub>
</nav>
`;

exports[`User Menu component when no quota and no email is set the user menu does not contain a quota 1`] = `
<nav aria-label="Account menu">
  <oc-button-stub type="button" size="medium" arialabel="User Menu" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium" id="_userMenuButton" class="oc-topbar-personal">
    <avatar-image-stub width="32" userid="einstein" class="oc-topbar-personal-avatar oc-flex-inline oc-flex-center oc-flex-middle"></avatar-image-stub>
  </oc-button-stub>
  <oc-drop-stub dropid="account-info-container" toggle="#_userMenuButton" position="bottom-start" mode="click" closeonclick="true" paddingsize="small">
    <oc-list-stub class="user-menu-list">
      <li>
        <oc-button-stub type="router-link" size="medium" to="[object Object]" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium" id="oc-topbar-account-manage">
          <avatar-image-stub width="32" userid="einstein"></avatar-image-stub> <span class="profile-info-wrapper oc-py-xs"><span></span>
          <!---->
          <!----></span>
        </oc-button-stub>
      </li>
      <li>
        <oc-button-stub type="router-link" size="medium" to="[object Object]" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium">
          <oc-icon-stub name="application" filltype="fill" accessiblelabel="" type="span" size="medium" variation="passive" color="" class="oc-p-xs"></oc-icon-stub>
          <translate-stub tag="span">Settings</translate-stub>
        </oc-button-stub>
      </li>
      <li>
        <oc-button-stub type="button" size="medium" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium" id="oc-topbar-account-logout">
          <oc-icon-stub name="logout-box-r" filltype="line" accessiblelabel="" type="span" size="medium" variation="passive" color="" class="oc-p-xs"></oc-icon-stub>
          <translate-stub tag="span">Log out</translate-stub>
        </oc-button-stub>
      </li>
      <!---->
    </oc-list-stub>
  </oc-drop-stub>
</nav>
`;

exports[`User Menu component when quota and no email is set renders a navigation without email 1`] = `
<nav aria-label="Account menu">
  <oc-button-stub type="button" size="medium" arialabel="User Menu" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium" id="_userMenuButton" class="oc-topbar-personal">
    <avatar-image-stub width="32" userid="einstein" class="oc-topbar-personal-avatar oc-flex-inline oc-flex-center oc-flex-middle"></avatar-image-stub>
  </oc-button-stub>
  <oc-drop-stub dropid="account-info-container" toggle="#_userMenuButton" position="bottom-start" mode="click" closeonclick="true" paddingsize="small">
    <oc-list-stub class="user-menu-list">
      <li>
        <oc-button-stub type="router-link" size="medium" to="[object Object]" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium" id="oc-topbar-account-manage">
          <avatar-image-stub width="32" userid="einstein"></avatar-image-stub> <span class="profile-info-wrapper oc-py-xs"><span></span>
          <!---->
          <!----></span>
        </oc-button-stub>
      </li>
      <li>
        <oc-button-stub type="router-link" size="medium" to="[object Object]" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium">
          <oc-icon-stub name="application" filltype="fill" accessiblelabel="" type="span" size="medium" variation="passive" color="" class="oc-p-xs"></oc-icon-stub>
          <translate-stub tag="span">Settings</translate-stub>
        </oc-button-stub>
      </li>
      <li>
        <oc-button-stub type="button" size="medium" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium" id="oc-topbar-account-logout">
          <oc-icon-stub name="logout-box-r" filltype="line" accessiblelabel="" type="span" size="medium" variation="passive" color="" class="oc-p-xs"></oc-icon-stub>
          <translate-stub tag="span">Log out</translate-stub>
        </oc-button-stub>
      </li>
      <li class="storage-wrapper oc-pl-s">
        <oc-icon-stub name="cloud" filltype="line" accessiblelabel="" type="span" size="medium" variation="passive" color="" class="oc-p-xs"></oc-icon-stub>
        <div class="storage-wrapper-text">
          <p class="oc-my-rm"><span>Personal storage (30.0% used)</span> <br> <span class="oc-text-small">300 B of 1 kB used</span></p>
          <oc-progress-stub value="30" max="100" size="small" variation="primary"></oc-progress-stub>
        </div>
      </li>
    </oc-list-stub>
  </oc-drop-stub>
</nav>
`;

exports[`User Menu component when quota is above 80% and below 90% renders a warning quota progress bar 1`] = `
<nav aria-label="Account menu">
  <oc-button-stub type="button" size="medium" arialabel="User Menu" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium" id="_userMenuButton" class="oc-topbar-personal">
    <avatar-image-stub width="32" userid="einstein" class="oc-topbar-personal-avatar oc-flex-inline oc-flex-center oc-flex-middle"></avatar-image-stub>
  </oc-button-stub>
  <oc-drop-stub dropid="account-info-container" toggle="#_userMenuButton" position="bottom-start" mode="click" closeonclick="true" paddingsize="small">
    <oc-list-stub class="user-menu-list">
      <li>
        <oc-button-stub type="router-link" size="medium" to="[object Object]" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium" id="oc-topbar-account-manage">
          <avatar-image-stub width="32" userid="einstein"></avatar-image-stub> <span class="profile-info-wrapper oc-py-xs"><span></span>
          <!---->
          <!----></span>
        </oc-button-stub>
      </li>
      <li>
        <oc-button-stub type="router-link" size="medium" to="[object Object]" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium">
          <oc-icon-stub name="application" filltype="fill" accessiblelabel="" type="span" size="medium" variation="passive" color="" class="oc-p-xs"></oc-icon-stub>
          <translate-stub tag="span">Settings</translate-stub>
        </oc-button-stub>
      </li>
      <li>
        <oc-button-stub type="button" size="medium" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium" id="oc-topbar-account-logout">
          <oc-icon-stub name="logout-box-r" filltype="line" accessiblelabel="" type="span" size="medium" variation="passive" color="" class="oc-p-xs"></oc-icon-stub>
          <translate-stub tag="span">Log out</translate-stub>
        </oc-button-stub>
      </li>
      <li class="storage-wrapper oc-pl-s">
        <oc-icon-stub name="cloud" filltype="line" accessiblelabel="" type="span" size="medium" variation="passive" color="" class="oc-p-xs"></oc-icon-stub>
        <div class="storage-wrapper-text">
          <p class="oc-my-rm"><span>Personal storage (81.0% used)</span> <br> <span class="oc-text-small">810 B of 1 kB used</span></p>
          <oc-progress-stub value="81" max="100" size="small" variation="warning"></oc-progress-stub>
        </div>
      </li>
    </oc-list-stub>
  </oc-drop-stub>
</nav>
`;

exports[`User Menu component when quota is below 80% renders a primary quota progress bar 1`] = `
<nav aria-label="Account menu">
  <oc-button-stub type="button" size="medium" arialabel="User Menu" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium" id="_userMenuButton" class="oc-topbar-personal">
    <avatar-image-stub width="32" userid="einstein" class="oc-topbar-personal-avatar oc-flex-inline oc-flex-center oc-flex-middle"></avatar-image-stub>
  </oc-button-stub>
  <oc-drop-stub dropid="account-info-container" toggle="#_userMenuButton" position="bottom-start" mode="click" closeonclick="true" paddingsize="small">
    <oc-list-stub class="user-menu-list">
      <li>
        <oc-button-stub type="router-link" size="medium" to="[object Object]" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium" id="oc-topbar-account-manage">
          <avatar-image-stub width="32" userid="einstein"></avatar-image-stub> <span class="profile-info-wrapper oc-py-xs"><span></span>
          <!---->
          <!----></span>
        </oc-button-stub>
      </li>
      <li>
        <oc-button-stub type="router-link" size="medium" to="[object Object]" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium">
          <oc-icon-stub name="application" filltype="fill" accessiblelabel="" type="span" size="medium" variation="passive" color="" class="oc-p-xs"></oc-icon-stub>
          <translate-stub tag="span">Settings</translate-stub>
        </oc-button-stub>
      </li>
      <li>
        <oc-button-stub type="button" size="medium" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium" id="oc-topbar-account-logout">
          <oc-icon-stub name="logout-box-r" filltype="line" accessiblelabel="" type="span" size="medium" variation="passive" color="" class="oc-p-xs"></oc-icon-stub>
          <translate-stub tag="span">Log out</translate-stub>
        </oc-button-stub>
      </li>
      <li class="storage-wrapper oc-pl-s">
        <oc-icon-stub name="cloud" filltype="line" accessiblelabel="" type="span" size="medium" variation="passive" color="" class="oc-p-xs"></oc-icon-stub>
        <div class="storage-wrapper-text">
          <p class="oc-my-rm"><span>Personal storage (30.0% used)</span> <br> <span class="oc-text-small">300 B of 1 kB used</span></p>
          <oc-progress-stub value="30" max="100" size="small" variation="primary"></oc-progress-stub>
        </div>
      </li>
    </oc-list-stub>
  </oc-drop-stub>
</nav>
`;
